src/CompositeRheologies.jl

```diff
@@ -0,0 +1,794 @@
+        τII = zero(_T)
+        Base.Cartesian.@nexprs $N i ->
+            τII += compute_τII(v.elements[i], εII, args)
+    end
+end
+compute_τII_AD(v::Parallel{T,N}, εII::_T, args; tol=1e-6, verbose=false) where {T,N,_T} = compute_τII(v, εII, args) 
+
+
+@generated  function compute_τII_i(
+    v::CompositeRheology{T,N}, 
+    εII::_T, 
+    args, I::Int64;
+    tol=1e-6, verbose=false
+) where {T,_T,N}
+    quote
+        Base.@_inline_meta
+        Base.Cartesian.@nexprs $N i -> I == i && return compute_τII(v.elements[i], εII, args)
+    end
+end
+
+function compute_τII_AD(v::CompositeRheology, εII, args; tol=1e-6, verbose=false)
+     τII = local_iterations_εII_AD(v, εII, args; tol=tol, verbose=verbose)
+     return τII
+end
+
+
+@inline function compute_τII!(
+    τII::AbstractArray{T,nDim},
+    v::NTuple{N,AbstractConstitutiveLaw},
+    εII::AbstractArray{T,nDim},
+    args;
+    tol=1e-6, verbose=false
+    ) where {T,nDim,N}
+    for I in eachindex(τII)
+        τII[I] = compute_τII(v, εII[I], (; zip(keys(args), getindex.(values(args), I))...))
+    end
+end
+
+# VISCOSITY COMPUTATIONS
+
+""" 
+    η = computeViscosity_εII(v::Union{Parallel{T,N}, CompositeRheology{T,N}, AbstractConstitutiveLaw}, εII::_T, args; tol=1e-6, verbose=false)
+
+This computes the effective viscosity for a given input rheology `v` and strainrate `εII`
+"""
+function computeViscosity_εII(
+    v::Union{Parallel{T,N}, CompositeRheology{T,N}}, 
+    εII::_T, 
+    args;
+    tol=1e-6, verbose=false
+) where {T,N,_T}
+    τII = compute_τII(v, εII, args; tol=tol, verbose=verbose)
+    η   = _T(0.5) * τII * inv(εII)
+    return η
+end
+
+function computeViscosity_εII(v::T, εII::_T, args; tol=1e-6, verbose=false) where {T<:AbstractConstitutiveLaw,_T}
+    τII = compute_τII(v, εII, args)
+    η   = _T(0.5) * τII * inv(εII)
+    return η
+end
+
+""" 
+    η = computeViscosity_εII_AD(v::Union{Parallel{T,N}, CompositeRheology{T,N}, AbstractConstitutiveLaw}, εII::_T, args; tol=1e-6, verbose=false)
+
+This computes the effective viscosity for a given input rheology `v` and strainrate `εII`, while using AD if necessary
+"""
+function computeViscosity_εII_AD(
+    v::Union{Parallel{T,N}, CompositeRheology{T,N}, AbstractConstitutiveLaw}, 
+    εII::_T, 
+    args;
+    tol=1e-6, verbose=false
+) where {T,N,_T}
+    τII = compute_τII_AD(v, εII, args; tol=tol, verbose=verbose)
+    η   = _T(0.5) * τII * inv(εII)
+    return η
+end
+
+function computeViscosity_εII_AD(v::T, εII::_T, args; tol=1e-6, verbose=false) where {T<:AbstractConstitutiveLaw,_T}
+    return computeViscosity_εII(v, εII, args) 
+end
+
+# NONLINEAR ITERATION SCHEMES
+"""
+    τII =local_iterations_εII_AD(v::CompositeRheology{N,T,0}, εII::_T, args; tol=1e-6, verbose=false)
+
+Performs local iterations versus stress for a given total strain rate for a given `CompositeRheology` element that does NOT include `Parallel` elements
+"""
+@inline function local_iterations_εII(
+    v::CompositeRheology{N,T,0}, 
```

Comment:
Again, is the type signature correct?

Reply:
as above

Reply:
actually, it was indeed incorrect (should be {T,N,0} or the full one)